Updated GiD problem type and parsers

MarkoLeskovar opened the pull request:
Dear Andreas,

I expanded the Structural and Fluid parser by field propFSI. It is added as the last argument in the parser and contains information about coupled boundary.

Best regards,
Marko Leskovar

MarkoLeskovar committed:
Updated GiD problem type and parsers

MarkoLeskovar committed:
updated FSI nodes names

andreas-apostolatos commented:
Dear Marko,

thank you very much for the enhancement of the parser. 

Can you add an example script under main/main_FEMComputationalFluidStructureInteraction where the parser is used on both the fluid and the structural domains, to test the motion transfer algorithm that I am implementing? 

It would be great if the input file would be that of the Turek FSI benchmark.

Thanks!
Andreas

> Dear Andreas,
> 
> I expanded the Structural and Fluid parser by field propFSI. It is added as the last argument in the parser and contains information about coupled boundary.
> 
> Best regards,
> Marko Leskovar

andreas-apostolatos commented:
> Dear Andreas,
> 
> I expanded the Structural and Fluid parser by field propFSI. It is added as the last argument in the parser and contains information about coupled boundary.
> 
> Best regards,
> Marko Leskovar

Dear Marko,

I answered to your previous message but I am not sure if you noticed it, as you wrote a new message simultaneously. 

Please take care of that first so that we can work in parallel.

Thanks,
Andreas

MarkoLeskovar commented:
Dear Andreas,

I will take care of it right now.

Best regards,
Marko

MarkoLeskovar committed:
Added test script that combines data from both parsers for Fluid and Structure model

andreas-apostolatos commented on parsers/parse_FluidModelFromGid.m:
```diff
@@ -232,6 +240,7 @@
 end
 out = cell2mat(out);
 fldMsh.nodes = out(:,2:4);
+fldMsh.nodeIDs = out(:,1);
```
Why do we need the node IDs from GiD? I guess that is because when we have both structural and fluid nodes in the GiD input file then the nodes of the fluid problem don't start from 1, right? I guess in that case we will have a problem running the CFD case itself as the code in the current state assumes that the nodes are sequentially numbered starting from 1, right?

andreas-apostolatos commented on parsers/parse_FluidModelFromGid.m:
```diff
@@ -252,6 +261,7 @@
 end
 out = cell2mat(out);
 fldMsh.elements = out(:, 2:end);
+fldMsh.elementIDs = out(:, 1);
```
The same discussion as for the nodes holds for the element numbering as well, right?

andreas-apostolatos reviewed (commented):
Hi Marko,

it seems a little bit problematic that we need to add the IDs for both the nodes and the elements coming from GiD. I guess that is a prerequisite if we want to read both the fluid and the structural meshes from the same GiD input file, right?

If that's the case I am afraid that both the structural and the fluid analysis will have to be adapted so that the nodal and element IDs are read in from GiD rather than being assumed sequential in the code.

What do you think?

Kind Regards,
Andreas

MarkoLeskovar commented:
Dear Andreas,

You are right regarding the GiD. The numbering of the mesh nodes in GiD is independent from the domains. Until now we had only one domain specified in GiD (structural, fluid, or thermal). Therefore, the numbering of nodes in the domain was the same as the numbering of the mesh nodes -> sequential numbering. But now we have two domains in GiD that share the same “global” mesh and “global” node numbering. GiD then outputs which nodes belong to which domain and the numbering is not sequential anymore. 

I think there is no need to further change the GiD problem type and parsers. Instead of passing in the fldMsh or strMsh in FSI problems you simply pass in the globalMsh and find corresponding elements and nodes from there. Existing problems can stay as they are. 
Example: strNodes = globalMsh.nodes(strMsh.nodeIDs,:);

Another way this could be done (I haven't tried it yet) is to completely change the GiD input file to have only "GLOBAL_NODES" and "GLOBAL_ELEMENTS" and then specify the numbering correponding to each domain. But this of course would require from us to change the parsers and all existing input files.

Best regards,
Marko Leskovar

andreas-apostolatos commented:
> Dear Andreas,
> 
> You are right regarding the GiD. The numbering of the mesh nodes in GiD is independent from the domains. Until now we had only one domain specified in GiD (structural, fluid, or thermal). Therefore, the numbering of nodes in the domain was the same as the numbering of the mesh nodes -> sequential numbering. But now we have two domains in GiD that share the same “global” mesh and “global” node numbering. GiD then outputs which nodes belong to which domain and the numbering is not sequential anymore.
> 
> I think there is no need to further change the GiD problem type and parsers. Instead of passing in the fldMsh or strMsh in FSI problems you simply pass in the globalMsh and find corresponding elements and nodes from there. Existing problems can stay as they are.
> Example: strNodes = globalMsh.nodes(strMsh.nodeIDs,:);
> 
> Another way this could be done (I haven't tried it yet) is to completely change the GiD input file to have only "GLOBAL_NODES" and "GLOBAL_ELEMENTS" and then specify the numbering correponding to each domain. But this of course would require from us to change the parsers and all existing input files.
> 
> Best regards,
> Marko Leskovar

Hi Marko,

thank you very much for the detailed answer, now it is clear.

I now remembered that this is a very big weakness of the code which I always wanted to fix but never allocated time for it. As you accurately noted, the problem is that, both the nodes and the elements are read-in by neglecting their IDs coming from GiD and assuming simply a sequential numbering for them starting from 1 within MATLAB.

The solution is to simply expand the array for the nodes and the elements to have in their fist column (:,1) there IDs provided by GiD stored. One has to go through every call of these .nodes and .elements and change the index. It really should not be that hard and in this way we can skip defining new arrays with the name IDs(:, 1), which I don't like as a solution. In this way, only the parsers (fluid, structure and thermal) as well as every call msh.nodes and msh.elements have to be changed whereas all input files, drivers scripts, wrappers and unit tests can stay as they are!

Can you please put this in your pipeline within this pull request?

Thank you so much for your outstanding work,
Andreas

MarkoLeskovar commented:
Dear Andreas,

This is exactly why I added a new array instead of changing the existing one. I think that is better to have it seperate and simply don't use it than to go and change every function that calls "nodes" and "elements" which I fell like it's 90% of our code :)

I am glad that you like my work.
Best regards,
Marko

andreas-apostolatos commented:
> Dear Andreas,
> 
> This is exactly why I added a new array instead of changing the existing one. I think that is better to have it seperate and simply don't use it than to go and change every function that calls "nodes" and "elements" which I fell like it's 90% of our code :)
> 
> I am glad that you like my work.
> Best regards,
> Marko

Hi Marko,

yes I understood that, However in this way there is an inconsistency being added in the code when introducing such a variable.

I would instead like to have it consistently defined everywhere in the code avoiding having to always take care of an extra variable. Additionally, I am convinced that it is not overly complicated to go and find all instances of .nodes() and .elements() all over the code and change them accordingly. It really should be just a little tedious work but relatively easy to do.

Can you please try and report in case this becomes complicated?

Thanks a lot for your efforts,
Andreas

MarkoLeskovar commented:
Dear Andreas,

well in that case I will go ahead and try to change it. 

Best reagrds,
Marko

MarkoLeskovar committed:
Updated mesh.elements and mesh.nodes structure

andreas-apostolatos commented on FEMPlateInMembraneActionAnalysis/solutionMatricesAndVectors/computeStiffnessMatrixPlateInMembraneActionLinear.m:
```diff
@@ -50,10 +50,10 @@
     
     % Get the current element in the mesh
-    element = mesh.elements(i,1:no_nodes_element);
+    element = mesh.elements(i,2:no_nodes_element);
```
Shouldn't that here be,

element = mesh.elements(i,2:no_nodes_element + 1);

? It seems peculiar in comparison to the rest of the changes ..

MarkoLeskovar replied:
Yes, you are right. I missed that. This is what I meant by "a file that I missed" :) I will check the changes again

andreas-apostolatos replied:
Ahh great I spotted that :) The easier would be to do a search in all files within the cane/ directory for **.elements** and **.nodes**, which will give you all instances you will have to change. Under linux you can try the command,

grep -r "\\.nodes"
grep -r "\\.elements"

I will perform the command and send you the outcome in a separate e-mail so that you can double check that all corresponding instances have been found

andreas-apostolatos reviewed (commented):
Dear Marko,

you are too fast :D great! Please have a look in that minor thing I observed and let me know if it is supposed to be like that.

I will start tonight with the coupling of the structural and fluid DOFs for the Turek benchmark.

Needless to say how much you've helped!

Greetings,
Andreas

MarkoLeskovar committed:
Mesh numbering conversion for FSI cases

MarkoLeskovar commented:
Dear Andreas,

I changed the code structure as you requested. I ran the unit tests and individual main files and it seems OK, but there maybe is still a file that I missed.

Futhermore, I solved the meshes in FSI test case. Global node numbering is now converted to individual node numbering for each mesh (structure and fluid). Both cases can be now run as individual files with individual mesh. Take a look at the ploting options at the end. Let me know what you think.

Best regards,
Marko

andreas-apostolatos commented:
> Dear Andreas,
> 
> I changed the code structure as you requested. I ran the unit tests and individual main files and it seems OK, but there maybe is still a file that I missed.
> 
> Futhermore, I solved the meshes in FSI test case. Global node numbering is now converted to individual node numbering for each mesh (structure and fluid). Both cases can be now run as individual files with individual mesh. Take a look at the ploting options at the end. Let me know what you think.
> 
> Best regards,
> Marko

Hi Marko,

Many thanks for the efforts.

What do you mean by saying that there maybe a file that still is not fixed? Do you know which file or you think abstractly that you may have missed one?

Moreover, please have a look at one line of code that I found peculiar, I mentioned it within my last review to this pull request.

Thank you!
Andreas

MarkoLeskovar committed:
Double-checked the previous changes related to "nodes" and "elements"

andreas-apostolatos reviewed (approved):
Dear Marko,

this is excellent, I am approving the pull request. I also added the missing function so now the unit test run again.

Thanks a lot,
Andreas
